### ISSUE TITLE
self.is_initializing and model.apply(..., mutable=True)

### ISSUE THREAD

**Issue by zhangqiaorjc** [google/flax#2232]
It seems a Flax idiom to define 

```
def setup(self):
    self.is_initializing = self.is_mutable_collection('params')
```

and use it in 
```
def __call__(self, x):
    if not self.is_initializing:
       # mutate batch stats or some other state
```

However, `self.is_mutable_collection('params')` doesn't seem like a bullet-proof way to find out if we are inside model.init, because users may want to do

`model.apply(..., mutable=True)` in their unit tests for example.

I wonder if there's a more robust way to tell if we are inside model.init vs. model.apply.

**Comment by zhangqiaorjc** [comment]
Perhaps we should avoid `mutable=True` with `model.apply` since it makes params mutable and they may get mutated by accident.

But in general is `self.is_initializing = self.is_mutable_collection('params')` the best way to tell that we are inside `model.init`? It seems indirect. 

Could nn.Module tag something and helps the users to avoid having to test if 'params' is mutable?

**Comment by cgarciae** [comment]
Hey @zhangqiaorjc! There is an ongoing discussion about this topic in #652 and #2131. The case you describe is very pathological because setting `mutable=True` often unintuitively triggers initialization logic which can result in e.g. batch statistics not updating.

@jheek brought an interesting idea of having "initializable collections", presumably you could get an API similar to:

```python
# initialization
module.init(..., initializable=True) # default value
# other `CollectionFilter`s would be valid e.g. ['params', 'batch_stats', ...]
...
# training
module.apply(..., initializable=False) # default value
...
# inside a Module
self.is_initializing = self.is_initializable_collection('params') 
```
I am very interested in this topic.

**Comment by andsteing** [comment]
See #2234

**Comment by levskaya** [comment]
Fixed by #2234 - (further comments always welcome).
